server-spi-private/src/main/java/org/keycloak/userprofile/UserProfileProvider.java

```diff
@@ -0,0 +1,35 @@
+/*
+ * Copyright 2020 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.keycloak.userprofile;
+
+import org.keycloak.userprofile.util.UpdateUserProfileContext;
+import org.keycloak.userprofile.validation.UserProfileValidationResult;
+import org.keycloak.provider.Provider;
+import org.keycloak.provider.ProviderFactory;
+import org.keycloak.representations.idm.UserAttributesRepresentation;
+import org.keycloak.userprofile.validation.ValidationScenario;
+
+/**
+ * @author <a href="mailto:markus.till@bosch.io">Markus Till</a>
+ */
+public interface UserProfileProvider extends Provider, ProviderFactory<UserProfileProvider> {
+
+    UserProfileValidationResult validate(UserAttributesRepresentation rep, UpdateUserProfileContext user, ValidationScenario validationScenario);
+
+    void update(UserAttributesRepresentation rep, UpdateUserProfileContext user, ValidationScenario validationScenario);
```

Comment:
What would you prefer instead? A UserUpdateProvider? To me, it was important to have a unique entry point for updating users as the old behavior was very error-prone (and yes I found "relevant" errors in the different code paths which are gone now). 
I would prefer a single stable update method. Rather in this provider or its own one. This also would enable profile implementations to execute actions on save e.g. external validation or data source checks. But may I'm wrong here.

Reply:
I agree it would be nice to have a single place that deals with updating users, we probably don't need an SPI for it. As simple way for now could be to just add a UserUpdateHelper or something, that can be refactored later as we are planning to clean up the stores quite a bit, especially around this sort of stuff. @hmlnarik wdyt?